src/bin/server.rs

```diff
@@ -8,15 +8,59 @@
 
 use mini_redis::{server, DEFAULT_PORT};
 
+#[cfg(feature = "xray")]
+use opentelemetry::global;
+#[cfg(feature = "xray")]
+use opentelemetry::sdk::trace as sdktrace;
+#[cfg(feature = "xray")]
+use opentelemetry_aws::trace::XrayPropagator;
 use structopt::StructOpt;
 use tokio::net::TcpListener;
 use tokio::signal;
+#[cfg(feature = "xray")]
+use tracing_subscriber::{layer::SubscriberExt, util::SubscriberInitExt};
 
 #[tokio::main]
 pub async fn main() -> mini_redis::Result<()> {
-    // enable logging
-    // see https://docs.rs/tracing for more info
-    tracing_subscriber::fmt::try_init()?;
+    #[cfg(feature = "xray")]
+    {
+        // Set the global propagator to X-Ray propagator
+        // Note: If you need to pass the x-amzn-trace-id across services in the same trace,
+        // you will need this line. However, this requires additional code not pictured here.
+        // For a full example using hyper, see:
+        // https://github.com/open-telemetry/opentelemetry-rust/blob/main/examples/aws-xray/src/server.rs#L14-L26
+        global::set_text_map_propagator(XrayPropagator::default());
+
+        let tracer = opentelemetry_otlp::new_pipeline()
+            .tracing()
+            .with_exporter(opentelemetry_otlp::new_exporter().tonic())
+            .with_trace_config(
+                sdktrace::config()
+                    .with_sampler(sdktrace::Sampler::AlwaysOn)
+                    // Needed in order to convert the trace IDs into an Xray-compatible format
+                    .with_id_generator(sdktrace::XrayIdGenerator::default()),
+            )
+            .install_simple()?;
+
+        // Create a tracing layer with the configured tracer
+        let opentelemetry = tracing_opentelemetry::layer().with_tracer(tracer);
+
+        let filter_layer = tracing_subscriber::EnvFilter::new("mini_redis=trace");
+
+        // Use the tracing subscriber `Registry`, or any other subscriber
+        // that impls `LookupSpan`
+        tracing_subscriber::registry()
+            .with(opentelemetry)
+            .with(filter_layer)
```

Comment:
it seems like it could be worth still including the `fmt` layer for logging to the console as well; otherwise, running `mini-redis` won't print anything out at all if xray is enabled

Reply:
Makes sense